File: src/product_register/converters/shopify.py
from __future__ import annotations

from collections import defaultdict

from product_register.converters.base import BaseConverter
from product_register.models import ProductInput

SHOPIFY_CDN_BASE = "https://cdn.shopify.com/s/files/1/0602/0992/2282/files/"

SITE_NAME = "【くりま】沖縄県産品・特産品の通販サイト"
SEO_DESC_SUFFIX = "のことなら沖縄県産品・特産品の通販サイトくりま。沖縄から全国一律9800円以上購入で送料無料！単品購入で送料無料も多数。"

# All column keys in the Shopify CSV output (60 columns, matches expected format)
SHOPIFY_COLUMNS = [
    "Handle",
    "Image Src",
    "Image Position",
    "Title",
    "Body (HTML)",
    "Vendor",
    "Product Category",
    "Type",
    "Tags",
    "Published",
    "Option1 Name",
    "Option1 Value",
    "Option1 Linked To",
    "Option2 Name",
    "Option2 Value",
    "Option2 Linked To",
    "Option3 Name",
    "Option3 Value",
    "Option3 Linked To",
    "Variant SKU",
    "Variant Grams",
    "Variant Inventory Tracker",
    "Variant Inventory Qty",
    "Variant Inventory Policy",
    "Variant Fulfillment Service",
    "Variant Price",
    "Variant Compare At Price",
    "Variant Requires Shipping",
    "Variant Taxable",
    "Variant Barcode",
    "Image Alt Text",
    "Gift Card",
    "SEO Title",
    "SEO Description",
    "Google Shopping / Google Product Category",
    "Google Shopping / Gender",
    "Google Shopping / Age Group",
    "Google Shopping / MPN",
    "Google Shopping / Condition",
    "Google Shopping / Custom Product",
    "Google Shopping / Custom Label 0",
    "Google Shopping / Custom Label 1",
    "Google Shopping / Custom Label 2",
    "Google Shopping / Custom Label 3",
    "Google Shopping / Custom Label 4",
    "Google: Custom Product (product.metafields.mm-google-shopping.custom_product)",
    "送料 (product.metafields.my_fields._0000)",
    "商品評価数 (product.metafields.reviews.rating_count)",
    "国 (product.metafields.shopify.country)",
    "Variant Image",
    "Variant Weight Unit",
    "Variant Tax Code",
    "Cost per item",
    "Included / 日本",
    "Price / 日本",
    "Compare At Price / 日本",
    "Included / 国際",
    "Price / 国際",
    "Compare At Price / 国際",
    "Status",
]


def _base_code(p: ProductInput) -> str:
    """Base code: maker_code + '-' + last 4 digits of jan_code"""
    return f"{p.maker_code}-{p.jan_code[-4:]}"


def _price_with_tax(p: ProductInput) -> int:
    """Tax-inclusive price: selling_price * (1 + tax_rate/100)"""
    return round(p.selling_price * (1 + p.tax_rate / 100))


def _generate_image_url(base: str, position: int) -> str:
    """Generate Shopify CDN image URL.

    Position 1: {base}-1.jpg
    Position 2+: {base}_{position}.jpg
    """
    if position == 1:
        return f"{SHOPIFY_CDN_BASE}{base}-1.jpg"
    return f"{SHOPIFY_CDN_BASE}{base}_{position}.jpg"


def _get_option1_value(p: ProductInput) -> str:
    """Determine Option1 Value for this product.

    The ne_code suffix (last segment after '-') is matched against variation_choices
    to find the applicable choice label.
    Append '(送料無料)' if shipping_type == '送料無料'.
    """
    if not p.variation_choices:
        return ""
    suffix = p.ne_code.rsplit("-", 1)[-1]
    choices = [c.strip() for c in p.variation_choices.split("|")]
    # Find the choice whose numeric digits match the suffix
    matched = None
    for choice in choices:
        digits = "".join(ch for ch in choice if ch.isdigit())
        if digits == suffix:
            matched = choice
            break
    if matched is None:
        matched = choices[0] if choices else ""
    if p.shipping_type == "送料無料":
        return f"{matched}(送料無料)"
    return matched


def _build_body_html(base: str, image_count: int, description_pc: str) -> str:
    """Build Body (HTML) with imgList section (images 2+) + description_pc."""
    img_tags = ""
    for i in range(2, image_count + 1):
        url = f"{SHOPIFY_CDN_BASE}{base}_{i}.jpg"
        img_tags += f'<img src="{url}" width="100%"><br>'
    img_list = f"<!--imgList-->{img_tags}<!--/imgList-->\n{description_pc}"
    return img_list


class ShopifyConverter(BaseConverter):
    mall_name = "shopify"

    def convert(self, products: list[ProductInput]) -> list[dict]:
        """ProductInputリストをShopify CSV用の辞書リストに変換する。

        商品はbase_code（maker_code + JAN下4桁）でグループ化される。
        各グループはmax(image_count)行に展開される。
        各行にはHandle + Image Src + Image Position が設定される。
        先頭行: 全フィールド設定（代表商品の情報）
        バリアント行: 対応するProductInputのバリアント情報 + 画像
        残り行: Handle + 画像のみ
        """
        # Group by base_code, preserving insertion order within each group
        groups: dict[str, list[ProductInput]] = defaultdict(list)
        for p in products:
            groups[_base_code(p)].append(p)

        rows: list[dict] = []
        for base in sorted(groups.keys()):
            members = groups[base]
            rows.extend(self._expand_group(base, members))
        return rows

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _expand_group(self, base: str, members: list[ProductInput]) -> list[dict]:
        """Expand a group of products sharing the same base code into rows."""
        # Use the first product (単品 preferred) as the representative for
        # product-level fields (Title, Body, Vendor, Tags, SEO, etc.)
        rep = next((p for p in members if p.is_single), members[0])
        max_images = max(p.image_count for p in members)

        rows: list[dict] = []
        for pos in range(1, max_images + 1):
            image_url = _generate_image_url(base, pos)

            if pos == 1:
                # First row: all product fields + first variant fields
                row = self._build_first_row(base, rep, members[0], image_url)
            elif pos - 1 < len(members):
                # Subsequent rows with a variant to attach
                variant = members[pos - 1]
                row = self._build_variant_row(base, variant, image_url, pos)
            else:
                # Image-only row
                row = self._build_image_only_row(base, image_url, pos)

            rows.append(row)
        return rows

    def _empty_row(self, handle: str) -> dict:
        """Return a blank row template with only Handle set."""
        row = {col: "" for col in SHOPIFY_COLUMNS}
        row["Handle"] = handle
        return row

    def _fill_variant_fields(self, row: dict, p: ProductInput, base: str) -> None:
        """Fill variant-specific fields into row in-place."""
        row["Option1 Value"] = _get_option1_value(p)
        row["Variant SKU"] = p.ne_code
        row["Variant Grams"] = "0.0"
        row["Variant Inventory Tracker"] = "shopify"
        row["Variant Inventory Qty"] = "0"
        row["Variant Inventory Policy"] = "deny"
        row["Variant Fulfillment Service"] = "manual"
        row["Variant Price"] = str(_price_with_tax(p))
        row["Variant Requires Shipping"] = "true"
        row["Variant Taxable"] = "true"
        row["Variant Barcode"] = p.jan_code
        # Variant Image: CDN URL using ne_code as filename
        row["Variant Image"] = f"{SHOPIFY_CDN_BASE}{p.ne_code}.jpg"

    def _build_first_row(
        self,
        base: str,
        rep: ProductInput,
        first_variant: ProductInput,
        image_url: str,
    ) -> dict:
        """Build the first row with all product + variant fields."""
        row = self._empty_row(base)

        # Product-level fields (from representative)
        row["Title"] = rep.display_name
        row["Body (HTML)"] = _build_body_html(base, rep.image_count, rep.description_pc)
        row["Vendor"] = f"{rep.maker_name}<br>商品コード:{base}"
        row["Tags"] = f"{rep.maker_name},税率{rep.tax_rate}%"
        row["Published"] = "true"

        # Option1 Name (use variation_name if available, else fixed label)
        if rep.variation_name:
            row["Option1 Name"] = "セット数を選んでください"
        else:
            row["Option1 Name"] = "Title"

        # SEO
        row["Gift Card"] = "false"
        row["SEO Title"] = f"{rep.product_name}|{SITE_NAME}"
        row["SEO Description"] = f"{rep.product_name}{SEO_DESC_SUFFIX}"

        # Included markets
        row["Included / 日本"] = "true"
        row["Included / 国際"] = "true"

        # Status
        row["Status"] = "active"

        # Image
        row["Image Src"] = image_url
        row["Image Position"] = "1"

        # Variant fields for the first variant
        self._fill_variant_fields(row, first_variant, base)
        row["Variant Weight Unit"] = "kg"

        return row

    def _build_variant_row(
        self,
        base: str,
        variant: ProductInput,
        image_url: str,
        position: int,
    ) -> dict:
        """Build a row for a subsequent variant (image + variant fields, no product-level fields)."""
        row = self._empty_row(base)
        row["Image Src"] = image_url
        row["Image Position"] = str(position)
        self._fill_variant_fields(row, variant, base)
        return row

    def _build_image_only_row(self, base: str, image_url: str, position: int) -> dict:
        """Build an image-only row (Handle + Image Src + Image Position)."""
        row = self._empty_row(base)
        row["Image Src"] = image_url
        row["Image Position"] = str(position)
        return row

File: webui/lib/converters/shopify.ts
import type { ProductInput, Variant } from "@/lib/product/schema";
import { productVariants } from "@/lib/product/schema";
import type { Converter } from "./base";
import { ENCODING } from "./base";

export const SHOPIFY_CDN_BASE = "https://cdn.shopify.com/s/files/1/0602/0992/2282/files/";
const SITE_NAME = "【くりま】沖縄県産品・特産品の通販サイト";
const SEO_DESC_SUFFIX =
  "のことなら沖縄県産品・特産品の通販サイトくりま。沖縄から全国一律9800円以上購入で送料無料！単品購入で送料無料も多数。";

export const SHOPIFY_COLUMNS = [
  "Handle", "Image Src", "Image Position", "Title", "Body (HTML)", "Vendor",
  "Product Category", "Type", "Tags", "Published",
  "Option1 Name", "Option1 Value", "Option1 Linked To",
  "Option2 Name", "Option2 Value", "Option2 Linked To",
  "Option3 Name", "Option3 Value", "Option3 Linked To",
  "Variant SKU", "Variant Grams", "Variant Inventory Tracker", "Variant Inventory Qty",
  "Variant Inventory Policy", "Variant Fulfillment Service", "Variant Price",
  "Variant Compare At Price", "Variant Requires Shipping", "Variant Taxable", "Variant Barcode",
  "Image Alt Text", "Gift Card", "SEO Title", "SEO Description",
  "Google Shopping / Google Product Category", "Google Shopping / Gender",
  "Google Shopping / Age Group", "Google Shopping / MPN", "Google Shopping / Condition",
  "Google Shopping / Custom Product",
  "Google Shopping / Custom Label 0", "Google Shopping / Custom Label 1",
  "Google Shopping / Custom Label 2", "Google Shopping / Custom Label 3",
  "Google Shopping / Custom Label 4",
  "Google: Custom Product (product.metafields.mm-google-shopping.custom_product)",
  "送料 (product.metafields.my_fields._0000)",
  "商品評価数 (product.metafields.reviews.rating_count)",
  "国 (product.metafields.shopify.country)",
  "Variant Image", "Variant Weight Unit", "Variant Tax Code", "Cost per item",
  "Included / 日本", "Price / 日本", "Compare At Price / 日本",
  "Included / 国際", "Price / 国際", "Compare At Price / 国際",
  "Status",
] as const;

function baseCodeOf(p: ProductInput): string {
  return `${p.maker_code}-${p.jan_code.slice(-4)}`;
}

function priceWithTax(p: { selling_price: number; tax_rate: number }): number {
  return Math.round(p.selling_price * (1 + p.tax_rate / 100));
}

/** 1 SKU 行の元データ。多SKU商品は variants[] を展開し、フラット商品は1件合成（後方互換）。 */
type SkuSlot = { p: ProductInput; v: Variant };

function generateImageUrl(base: string, position: number): string {
  if (position === 1) return `${SHOPIFY_CDN_BASE}${base}-1.jpg`;
  return `${SHOPIFY_CDN_BASE}${base}_${position}.jpg`;
}

function getOption1Value(p: ProductInput): string {
  if (!p.variation_choices) return "";
  const suffix = p.ne_code.split("-").slice(-1)[0];
  const choices = p.variation_choices.split("|").map((c) => c.trim());
  let matched: string | null = null;
  for (const choice of choices) {
    const digits = choice.replace(/[^0-9]/g, "");
    if (digits === suffix) {
      matched = choice;
      break;
    }
  }
  if (matched === null) matched = choices[0] || "";
  if (p.shipping_type === "送料無料") return `${matched}(送料無料)`;
  return matched;
}

function addTableClass(html: string): string {
  return html.replace(/<table\s[^>]*>/g, (tag) => {
    if (tag.includes("class=")) return tag;
    return tag.replace("<table ", '<table class="itemtable" ');
  });
}

export function buildShopifyBodyHtml(base: string, imageCount: number, descriptionPc: string): string {
  let imgTags = "";
  for (let i = 2; i <= imageCount; i++) {
    imgTags += `<img src="${SHOPIFY_CDN_BASE}${base}_${i}.jpg" width="100%"><br>`;
  }
  const processed = addTableClass(descriptionPc);
  return `<!--imgList-->${imgTags}<!--/imgList-->\n${processed}`;
}

export class ShopifyConverter implements Converter {
  mallName = "shopify" as const;
  encoding = ENCODING.shopify;

  convert(products: ProductInput[]): Record<string, string>[] {
    const groups = new Map<string, ProductInput[]>();
    for (const p of products) {
      const b = baseCodeOf(p);
      if (!groups.has(b)) groups.set(b, []);
      groups.get(b)!.push(p);
    }

    const rows: Record<string, string>[] = [];
    for (const base of Array.from(groups.keys()).sort()) {
      const members = groups.get(base)!;
      rows.push(...this.expandGroup(base, members));
    }
    return rows;
  }

  private expandGroup(base: string, members: ProductInput[]): Record<string, string>[] {
    const rep = members.find((p) => p.is_single) ?? members[0];
    // 各商品の variants[] を SKU スロット列に展開（フラット商品は1件合成 → 従来の members と同数）。
    const skus: SkuSlot[] = members.flatMap((p) => productVariants(p).map((v) => ({ p, v })));
    const maxImages = Math.max(...members.map((p) => p.image_count));
    // SKU 数が画像数を超えても SKU 行を落とさない。
    const rowCount = Math.max(maxImages, skus.length);
    const rows: Record<string, string>[] = [];

    for (let pos = 1; pos <= rowCount; pos++) {
      const imageUrl = pos <= maxImages ? generateImageUrl(base, pos) : "";
      if (pos === 1) {
        rows.push(this.buildFirstRow(base, rep, skus[0], imageUrl, skus.length));
      } else if (pos - 1 < skus.length) {
        rows.push(this.buildVariantRow(base, skus[pos - 1], imageUrl, pos));
      } else {
        rows.push(this.buildImageOnlyRow(base, imageUrl, pos));
      }
    }
    return rows;
  }

  private emptyRow(handle: string): Record<string, string> {
    const row: Record<string, string> = Object.fromEntries(SHOPIFY_COLUMNS.map((c) => [c, ""]));
    row["Handle"] = handle;
    return row;
  }

  private fillVariantFields(row: Record<string, string>, { p, v }: SkuSlot): void {
    // variation_value を持つ variant はそれを選択肢ラベルに（送料無料は接尾辞付き）。
    // フラット商品の合成 variant は variation_value 空 → 従来の getOption1Value(p) にフォールバック。
    row["Option1 Value"] = v.variation_value
      ? v.shipping_type === "送料無料"
        ? `${v.variation_value}(送料無料)`
        : v.variation_value
      : getOption1Value(p);
    row["Variant SKU"] = v.ne_code;
    row["Variant Grams"] = "0.0";
    row["Variant Inventory Tracker"] = "shopify";
    row["Variant Inventory Qty"] = "0";
    row["Variant Inventory Policy"] = "deny";
    row["Variant Fulfillment Service"] = "manual";
    row["Variant Price"] = String(priceWithTax(v));
    row["Variant Requires Shipping"] = "true";
    row["Variant Taxable"] = "true";
    row["Variant Barcode"] = v.jan_code;
    row["Variant Image"] = `${SHOPIFY_CDN_BASE}${v.ne_code}.jpg`;
  }

  private buildFirstRow(
    base: string,
    rep: ProductInput,
    firstVariant: SkuSlot,
    imageUrl: string,
    skuCount: number,
  ): Record<string, string> {
    const row = this.emptyRow(base);
    row["Title"] = rep.display_name;
    row["Body (HTML)"] = buildShopifyBodyHtml(base, rep.image_count, rep.description_pc);
    row["Vendor"] = `${rep.maker_name}<br>商品コード:${base}`;
    row["Tags"] = `${rep.maker_name},税率${rep.tax_rate}%`;
    row["Published"] = "true";
    // 多SKU(2件以上)は variation_name 未定義でもセレクタ名を出す（Shopify は複数variantにオプション名必須）。
    row["Option1 Name"] = rep.variation_name || skuCount > 1 ? "セット数を選んでください" : "Title";
    row["Gift Card"] = "false";
    row["SEO Title"] = `${rep.product_name}|${SITE_NAME}`;
    row["SEO Description"] = `${rep.product_name}${SEO_DESC_SUFFIX}`;
    row["Included / 日本"] = "true";
    row["Included / 国際"] = "true";
    row["Status"] = "active";
    row["Image Src"] = imageUrl;
    row["Image Position"] = "1";
    this.fillVariantFields(row, firstVariant);
    row["Variant Weight Unit"] = "kg";
    return row;
  }

  private buildVariantRow(
    base: string,
    sku: SkuSlot,
    imageUrl: string,
    position: number,
  ): Record<string, string> {
    const row = this.emptyRow(base);
    row["Image Src"] = imageUrl;
    row["Image Position"] = imageUrl ? String(position) : "";
    this.fillVariantFields(row, sku);
    return row;
  }

  private buildImageOnlyRow(
    base: string,
    imageUrl: string,
    position: number,
  ): Record<string, string> {
    const row = this.emptyRow(base);
    row["Image Src"] = imageUrl;
    row["Image Position"] = String(position);
    return row;
  }
}
